sqlmesh/core/engine_adapter/shared.py

```diff
@@ -47,3 +55,74 @@ class DataObject(PydanticModel):
     schema_name: str = Field(alias="schema")
     name: str
     type: DataObjectType
+
+
+def set_catalog(
+    *,
+    override: t.Optional[CatalogSupport] = None,
+) -> t.Callable:
+    def decorator(func: t.Callable) -> t.Callable:
+        @functools.wraps(func)
+        def wrapper(*args: t.Any, **kwargs: t.Any) -> t.Any:
+            def _get_args_pos_and_kwarg_name(
+                func: t.Callable,
+            ) -> t.Optional[t.Tuple[str, int, str]]:
+                spec = inspect.getfullargspec(func)
+                for i, name in enumerate(spec.args):
+                    obj_type = spec.annotations.get(name)
+                    if obj_type == "SchemaName":
+                        return name, i, obj_type
+                    if obj_type == "TableName":
+                        return name, i, obj_type
+                return None
+
+            # Need to convert args to list in order to later do assignment to the object
+            list_args = list(args)
+            engine_adapter = list_args[0]
+            catalog_support = override or engine_adapter.CATALOG_SUPPORT
+            # If there is full catalog support then we have nothing to do
+            if catalog_support.is_full_support:
+                return func(*list_args, **kwargs)
+
+            # Get the field value and the container which it came from so we can update it later
+            location = _get_args_pos_and_kwarg_name(func)
+            if location is None:
+                return func(*list_args, **kwargs)
+            name, pos, obj_type = location
+            obj: t.Union[str, exp.Table]
+            container: t.Union[t.Dict, t.List]
+            key: t.Union[int, str]
+            obj, container, key = (
+                (kwargs.get(name), kwargs, name)  # type: ignore
+                if kwargs.get(name)
+                else (list_args[pos], list_args, pos)
+            )
+            # mypy shenanigans: https://github.com/python/mypy/issues/10740
+            to_table_func: t.Callable[[t.Union[str, exp.Table]], exp.Table] = exp.to_table
+            to_schema_func: t.Callable[[t.Union[str, exp.Table]], exp.Table] = to_schema
+            to_expression_func = to_table_func if obj_type == "TableName" else to_schema_func
+            expression = to_expression_func(obj.copy() if isinstance(obj, exp.Table) else obj)
+            catalog_name = expression.args.get("catalog")
+            if not catalog_name:
+                return func(*list_args, **kwargs)
+            # If we have a catalog and this engine doesn't support catalogs then we need to error
+            if catalog_support.is_unsupported:
+                raise UnsupportedCatalogOperationError(
+                    f"{engine_adapter.dialect} does not support catalogs and a catalog was provided: {catalog_name}"
+                )
+            # Remove the catalog name from the argument so the engine adapter doesn't try to use it
+            expression.set("catalog", None)
```

Comment:
why is this necessary? to_schema should handle it

Reply:
`to_schema`'s purpose is to create a table object without a table name but with `db` and `catalog` populated (if defined).

The goal of this line is to take whatever the expression object is and ensure it does not have catalog set. We do this because we are setting the catalog using the engine adapter and therefore we don't want it include in future SQL statements.